### ISSUE TITLE
Cannot read properties of undefined (reading 'data')

### ISSUE THREAD

**Issue by leandro-branco** [nrwl/nx#17450]
### Current Behavior

When using `pnpm` as package manager and trying to build a project that contains `"generatePackageJson": true` option, build target is failing. The same doesn't occur when using 'yarn'.

- Nx version `16.3.2`

### Expected Behavior

It should build 

### GitHub Repo

https://github.com/leandro-branco/nx-bug-report

### Steps to Reproduce

1. run `pnpm install`
2. run `nx run nest-app:build:production --verbose `

### Nx Report

```shell
❯ nx report                                 

 >  NX   Report complete - copy this into the issue template

   Node   : 16.16.0
   OS     : darwin x64
   pnpm   : 8.6.1
   Hasher : Native
   
   nx (global)        : 16.3.2
   nx                 : 16.3.2
   @nx/js             : 16.3.2
   @nx/jest           : 16.3.2
   @nx/linter         : 16.3.2
   @nx/workspace      : 16.3.2
   @nx/angular        : 16.3.2
   @nx/cypress        : 16.3.2
   @nx/eslint-plugin  : 16.3.2
   @nx/nest           : 16.3.2
   @nx/node           : 16.3.2
   @nx/web            : 16.3.2
   @nx/webpack        : 16.3.2
   typescript         : 5.0.2
```

### Failure Logs

```shell
HookWebpackError: Cannot read properties of undefined (reading 'data')
    at makeWebpackError (/Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/webpack@5.85.1/node_modules/webpack/lib/HookWebpackError.js:48:9)
    at /Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/webpack@5.85.1/node_modules/webpack/lib/Compilation.js:3068:12
    at eval (eval at create (/Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:23:1)
    at fn (/Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/webpack@5.85.1/node_modules/webpack/lib/Compilation.js:481:17)
    at _next0 (eval at create (/Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:21:1)
    at eval (eval at create (/Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:34:1)
    at eval (eval at create (/Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:11:1)
    at /Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/webpack@5.85.1/node_modules/webpack/lib/Compilation.js:503:10
    at Hook.eval [as callAsync] (eval at create (/Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:7:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/Hook.js:18:14)
-- inner error --
TypeError: Cannot read properties of undefined (reading 'data')
    at /Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/nx@16.3.2/node_modules/nx/src/plugins/js/package-json/create-package-json.js:36:55
    at Array.forEach (<anonymous>)
    at createPackageJson (/Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/nx@16.3.2/node_modules/nx/src/plugins/js/package-json/create-package-json.js:34:79)
    at /Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/@nx+webpack@16.3.2_@types+node@18.7.1_nx@16.3.2_typescript@5.0.2/node_modules/@nx/webpack/src/plugins/generate-package-json-plugin.js:34:64
    at fn (/Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/webpack@5.85.1/node_modules/webpack/lib/Compilation.js:479:10)
    at _next0 (eval at create (/Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:21:1)
    at eval (eval at create (/Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:34:1)
    at eval (eval at create (/Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:11:1)
    at /Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/webpack@5.85.1/node_modules/webpack/lib/Compilation.js:503:10
    at Hook.eval [as callAsync] (eval at create (/Users/github.com/leandro-branco/nx-fail/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:7:1)
```

### Operating System

- [X] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

Debugging the problem, I've just realized that when using `yarn`, the [create-package-json.ts at line 63](https://github.com/nrwl/nx/blob/e693642c35ca04d247cafeab229e8890917469b0/packages/nx/src/plugins/js/package-json/create-package-json.ts#L63) is capable of finding the `graph.externalNodes[dep]` with dep equals to 'npm:tslib':

```typescript
console.log(dep) // = npm:tslib
console.log(graph.externalNodes[dep]) // = 
// graph.externalNodes[dep]: {
//  type: 'npm',
//  name: 'npm:tslib',
//  data: {
//    version: '2.5.3',
//    packageName: 'tslib',
//    hash: 'sha512-mSxlJJwl3BMEQCUNnxXBU9jP4JBktcEGhURcPR6VQVlnP0FdDEsIaz0C35dXNGLyRfrATNofF0F5p2KPxQgB+w=='
//  }
// }
```

However, using `pnpm`, graph.externalNodes[dep] is undefined as the graph.externalNodes has no entry `'npm:tslib'` although it has:
```typescript
  'npm:tslib@1.14.': {
    type: 'npm',
    name: 'npm:tslib@1.14.',
    data: {
      version: 'npm:@slib@1.14.1',
      packageName: 'tslib@1.14.',
      hash: 'sha512-Xni35NKzjgMrwevysHTCArtLDpPvye8zV/0E4EyYn43P7/7qvQwPh9BGkHewbMulVntbigmcT7rdX3BNo9wRJg=='
    }
  },
  'npm:tslib@2.5.@npm:@slib@2.5.0': {
    type: 'npm',
    name: 'npm:tslib@2.5.@npm:@slib@2.5.0',
    data: {
      version: 'npm:@slib@2.5.0',
      packageName: 'tslib@2.5.',
      hash: 'sha512-336iVw3rtn2BUK7ORdIAHTyxHGRIHVReokCR3XjbckJMK7ms8FysBfhLR8IXnAgy7T0PTPNBWKiH514FOW/WSg=='
    }
  },
  'npm:tslib@2.5.@npm:@slib@2.5.2': {
    type: 'npm',
    name: 'npm:tslib@2.5.@npm:@slib@2.5.2',
    data: {
      version: 'npm:@slib@2.5.2',
      packageName: 'tslib@2.5.',
      hash: 'sha512-5svOrSA2w3iGFDs1HibEVBGbDrAY82bFQ3HZ3ixB+88nsbsWQoKqDRb5UBYAUPEzbBn6dAp5gRNXglySbx1MlA=='
    }
  },
  'npm:tslib@2.5.@npm:@slib@2.5.3': {
    type: 'npm',
    name: 'npm:tslib@2.5.@npm:@slib@2.5.3',
    data: {
      version: 'npm:@slib@2.5.3',
      packageName: 'tslib@2.5.',
      hash: 'sha512-mSxlJJwl3BMEQCUNnxXBU9jP4JBktcEGhURcPR6VQVlnP0FdDEsIaz0C35dXNGLyRfrATNofF0F5p2KPxQgB+w=='
    }
  },
```

**Comment by Tirke** [comment]
Experiencing the same, thanks for filling the issue. Reverted temporarly.

**Comment by joshuaja** [comment]
Confirming I'm experiencing the same thing. The build is completely broken. @Tirke - which package did you revert, `@nx.webpack`?

**Comment by Tirke** [comment]
@joshuaja I always migrate with `nx migrate` so I rollback the full migration of all packages to the previous version. In our case it's `16.2.2`.

**Comment by rassawyer** [comment]
After downgrading to nx 16.1.4, I get a different error:
```
 pnpm api --verbose

> integro@0.0.3 api /home/rassawyer/Code/Integro/ClickUp
> nx run api:serve "--verbose"


> nx run api:serve


 >  NX   RunStart


{
  "meta": {
    "nxCloudVersion": "16.0.5"
  },
  "branch": null,
  "runGroup": "b5a795cd4ae308686747e34f228f53b0e9ba67ce",
  "ciExecutionId": null,
  "ciExecutionEnv": "",
  "hashes": [
    "4ffe385e683d72599c41d49b885e7bed763a53066980a581a19d123521be7d12"
  ]
}

RunStart duration: 392
Watch error: Daemon closed the connection

 ——————————————————————————————————————————————————————————————————————————————————

 >  NX   Running target serve for project api failed

   Failed tasks:

   - api:serve

   Hint: run the command with --verbose for more details.

   View structured, searchable error logs at https://nx.app/runs/EZiOXCyFmt

 ELIFECYCLE  Command failed with exit code 1.

```

**Comment by rassawyer** [comment]
If I add `"useDaemonProcess": "false"` to the runner options in nx.json, I get this:
```
 pnpm api --verbose

> integro@0.0.3 api /home/rassawyer/Code/Integro/ClickUp
> nx run api:serve "--verbose"


> nx run api:serve


 >  NX   RunStart


{
  "meta": {
    "nxCloudVersion": "16.0.5"
  },
  "branch": null,
  "runGroup": "b5a795cd4ae308686747e34f228f53b0e9ba67ce",
  "ciExecutionId": null,
  "ciExecutionEnv": "",
  "hashes": [
    "7bda8623da5b7d0ce3dfc843ebf828ed2b0961f589d11cd640c7a71e1cc5e541"
  ]
}

RunStart duration: 335

 >  NX   RunStart


{
  "meta": {
    "nxCloudVersion": "16.0.5"
  },
  "branch": null,
  "runGroup": "b5a795cd4ae308686747e34f228f53b0e9ba67ce",
  "ciExecutionId": null,
  "ciExecutionEnv": "",
  "hashes": [
    "0b085caff5109aac6df432602e6867efef2e7716ca8350d92bf38e1830685774"
  ]
}

RunStart duration: 1224

 >  NX   Nx Cloud: Cache miss 0b085caff5109aac6df432602e6867efef2e7716ca8350d92bf38e1830685774.


> nx run api:build

 >  NX   An error occured while creating pruned lockfile
   Please open an issue at `https://github.com/nrwl/nx/issues/new?template=1-bug.yml` and provide a reproduction.
   To prevent the build from breaking we are returning the root lock file.
   If you run `pnpm install --lockfile-only` in your output folder it will regenerate the correct pruned lockfile.

Original error: Pruned lock file creation failed. The following package was not found in the root lock file: @nestjs/common@9.4.2
   Error: Pruned lock file creation failed. The following package was not found in the root lock file: @nestjs/common@9.4.2
    at /home/rassawyer/Code/Integro/ClickUp/node_modules/.pnpm/nx@16.1.4/node_modules/nx/src/plugins/js/lock-file/project-graph-pruning.js:40:19
    at Array.forEach (<anonymous>)
    at normalizeDependencies (/home/rassawyer/Code/Integro/ClickUp/node_modules/.pnpm/nx@16.1.4/node_modules/nx/src/plugins/js/lock-file/project-graph-pruning.js:26:42)
    at pruneProjectGraph (/home/rassawyer/Code/Integro/ClickUp/node_modules/.pnpm/nx@16.1.4/node_modules/nx/src/plugins/js/lock-file/project-graph-pruning.js:13:34)
    at createLockFile (/home/rassawyer/Code/Integro/ClickUp/node_modules/.pnpm/nx@16.1.4/node_modules/nx/src/plugins/js/lock-file/lock-file.js:137:79)
    at /home/rassawyer/Code/Integro/ClickUp/node_modules/.pnpm/@nx+webpack@16.2.2_@types+node@18.7.1_nx@16.1.4_typescript@5.0.4/node_modules/@nx/webpack/src/plugins/generate-package-json-plugin.js:42:124
    at fn (/home/rassawyer/Code/Integro/ClickUp/node_modules/.pnpm/webpack@5.85.1/node_modules/webpack/lib/Compilation.js:479:10)
    at _next0 (eval at create (/home/rassawyer/Code/Integro/ClickUp/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:21:1)
    at eval (eval at create (/home/rassawyer/Code/Integro/ClickUp/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:34:1)
    at eval (eval at create (/home/rassawyer/Code/Integro/ClickUp/node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:11:1)
assets by path *.yaml 684 KiB
  asset pnpm-lock.yaml 683 KiB [emitted] [big]
  asset prod.yaml 520 bytes [emitted] [from: src/prod.yaml] [copied]
asset main.js 38.5 KiB [emitted] (name: main) 1 related asset
asset package.json 699 bytes [emitted]
asset assets/.gitkeep 0 bytes [emitted] [from: src/assets/.gitkeep] [copied]

ERROR in ./apps/api/src/keap/keap.service.ts:4:21
TS7016: Could not find a declaration file for module 'qs'. '/home/rassawyer/Code/Integro/ClickUp/node_modules/.pnpm/qs@6.11.2/node_modules/qs/lib/index.js' implicitly has an 'any' type.
  If the 'qs' package actually exposes this module, consider sending a pull request to amend 'https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/qs'
    2 | import { HttpException, Injectable } from "@nestjs/common"
    3 | import axios from "axios"
  > 4 | import * as qs from "qs"
      |                     ^^^^
    5 | import { reportErr } from "../shared/devops-email"
    6 | import { fireorm } from "../shared/fireorm"
    7 | import {
ERROR in ./apps/api/src/main.ts:9:22
TS7016: Could not find a declaration file for module 'body-parser'. '/home/rassawyer/Code/Integro/ClickUp/node_modules/.pnpm/body-parser@1.20.2/node_modules/body-parser/index.js' implicitly has an 'any' type.
  If the 'body-parser' package actually exposes this module, consider sending a pull request to amend 'https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/body-parser'
     7 | import { NestExpressApplication } from "@nestjs/platform-express"
     8 | import { apiEnv } from "./environments/api-environment"
  >  9 | import { json } from "body-parser"
       |                      ^^^^^^^^^^^^^
    10 | import { AllExceptionsFilter } from "./shared/exceptions-filter"
    11 | import * as helmet from "helmet"
    12 | import { reportErr } from "./shared/devops-email"
webpack 5.85.1 compiled with 2 errors (3dc988c5ceb8f430)



 >  NX   Running target build for project api failed

   Failed tasks:

   - api:build

   Hint: run the command with --verbose for more details.


 >  NX   RunEnd. Completed tasks


[
  {
    "taskId": "api:build",
    "startTime": "2023-06-13T11:48:13.220Z",
    "target": "build",
    "projectName": "api",
    "hash": "0b085caff5109aac6df432602e6867efef2e7716ca8350d92bf38e1830685774",
    "hashDetails": {
      "nodes": {
        "api:{projectRoot}/**/*": "b067b4837c4b853e522c58ca87042d9cc29035f1817459c0fdaf6e834e4b70e2",
        "{workspaceRoot}/nx.json": "6a134edd200955a5afa576dd835bbc6803a1642462083e9cf12d7842650625ed",
        "{workspaceRoot}/.gitignore": "9cd515fcdee64467410854eb36e25ef78774cd665e63cf654b18e58d87693b33",
        "{workspaceRoot}/.nxignore": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
        "models:{projectRoot}/**/*": "a81579b4042d69304e4cbdf085924c06ffe1fcc15120934ba5237b70da15bdbf",
        "utils:{projectRoot}/**/*": "5e9cdd858e31742d27c69c61d5d450e3d74d2f379c13756b5348c21b13a865b8"
      },
      "runtime": {},
      "implicitDeps": {}
    },
    "params": "",
    "uploadedToStorage": false,
    "endTime": "2023-06-13T11:48:17.890Z",
    "status": 1,
    "cacheStatus": "cache-miss",
    "terminalOutput": "\n \u001b[31m>\u001b[39m \u001b[0m\u001b[..."
  }
]

RunEnd duration: 299

Build failed, waiting for changes to restart...
```

**Comment by meeroslav** [comment]
This has been fixed in #17407
